std/traits.d

```diff
@@ -8590,3 +8591,83 @@ enum isCopyable(S) = is(typeof(
     static assert(isCopyable!int);
     static assert(isCopyable!(int[]));
 }
+
+/++
+ + Determines if `F` is a strongly `pure` function. This is true when :
+ + $(UL
+ +   $(LI Each parameter is `immutable`, or can be converted automatically to `immutable`
+ +   (i.e. has no mutable indirections) and is passed by value)
+ +   $(LI The Return Type is `immutable` or can be impicitly converted to `immutable`)
+ + )
+ + Params:
+ + F = The function to be tested
+ +
+ + Returns:
+ + `true` if the function is strongly `pure`, `false` if otherwise.
+ + ++/
+template isStronglyPure(F...)
+if (F.length == 1 && isCallable!F)
+{
+    bool isStronglyPure()
+    {
+        alias STC = ParameterStorageClass;
+
+        if (!hasFunctionAttributes!(F, "pure"))
+            return false;
+
+        static foreach (i, param; Parameters!F)
+        {
+            static if (is(param == immutable))
+            {}
+            else static if (isRefType!param || hasAliasing!param
+                    || ParameterStorageClassTuple!F[i] == STC.ref_
+                    || ParameterStorageClassTuple!F[i] == STC.out_)
+                return false;
+        }
+
+        if (!is(ReturnType!F : immutable(ReturnType!F)))
+            return false;
+
+        return true;
+    }
+
+    private bool isRefType(T)()
+    {
+        if (is(T == class) || isDelegate!T || isAssociativeArray!T || isPointer!T)
+            return true;
+
+        return false;
+    }
+}
+
+///
+@safe pure unittest
+{
+    int f(int x) pure
+    {
+        return x + 1;
+    }
+
+    void g(immutable ref int) pure
+    {}
+
+    void h(out int x) pure
+    {}
+
+    struct Test
+    {
+        int[string] aa;
+    }
+
+    void i(Test) pure
+    {}
+
+    void j(immutable (void function(ref int) pure)) pure
+    {}
+
+    static assert( isStronglyPure!f);
+    static assert( isStronglyPure!g);
+    static assert(!isStronglyPure!h);
+    static assert(!isStronglyPure!i); //has parameter Test with mutable reference member aa
+    static assert( isStronglyPure!j); //can call a function which is not strongly pure
```

Comment:
FYI: this is the publicly displayed example and while it does show off must of the functionality we have the informal consensus that public example snippets should ideally be 10 lines or less, because that's what a lazy reader can grasp within "one mind read".

-> how about splitting it up?

(you can also of course add more tests that aren't public or "hide" some of these)

Reply:
Done